interop-cats-effect/shared/src/main/scala/scalaz/zio/interop/bio/Concurrent2.scala

```diff
@@ -0,0 +1,334 @@
+
+  /**
+   * Returns an effect that races `fa1` with `fa2` returning the first
+   * successful `A` from the faster of the two. If one effect succeeds
+   * the other will be interrupted. If neither succeeds, then the
+   * effect will fail with some error.
+   *
+   * TODO: Example:
+   * {{{
+   *
+   * }}}
+   *
+   */
+  @inline def race[E, EE >: E, A, AA >: A](fa1: F[E, A], fa2: F[EE, AA])(
+    implicit CD: ConcurrentData2[F]
+  ): F[EE, Option[AA]] =
+    monad.map(raceEither(fa1, fa2))(_ map (_.merge))
+
+  /**
+   * Returns an effect that races `fa1` with `fa2`, calling specified
+   * finisher as soon as one result or the other has been computed.
+   *
+   * TODO: Example:
+   * {{{
+   *
+   * }}}
+   *
+   */
+  @inline def raceEither[E, EE >: E, A, B](fa1: F[E, A], fa2: F[EE, B])(
+    implicit ev: ConcurrentData2[F]
+  ): F[EE, Option[Either[A, B]]] = {
+
+    import cats.syntax.either._
+
+    implicit val _: Errorful2[F] = self
+
+    raceWith(fa1, fa2)(
+      (fa1Result, fiber2) =>
+        fa1Result.fold[F[EE, Option[Either[A, B]]]](monad.pure(None)) {
+          _.fold(
+            _ => fiber2.join map (f2r => Some(f2r.asRight)),
+            a => monad.pure(Some(a.asLeft)) <* fiber2.cancel
+          )
+        },
+      (fa2Result, fiber1) =>
+        fa2Result.fold[F[EE, Option[Either[A, B]]]](monad.pure(None)) {
+          _.fold(
+            _ => fiber1.join map (f1r => Some(f1r.asLeft)),
+            b => monad.pure(Some(b.asRight)) <* fiber1.cancel
+          )
+        }
+    )
+  }
+
+  /**
+   * Returns an effect that races `fa1` with `fa2`, calling specified
+   * finisher as soon as one result or the other has been computed.
+   *
+   * TODO: Example:
+   * {{{
+   *
+   * }}}
+   *
+   */
+  @inline def raceWith[E1, E2, E3, A, B, C](fa1: F[E1, A], fa2: F[E2, B])(
+    leftDone: (Option[Either[E1, A]], Fiber2[F, E2, B]) => F[E3, C],
+    rightDone: (Option[Either[E2, B]], Fiber2[F, E1, A]) => F[E3, C]
+  )(
+    implicit CD: ConcurrentData2[F]
+  ): F[E3, C] = {
+
+    implicit val ev: Errorful2[F] = self
+
+    def arbiter[EE0, EE1, AA, BB](
+      f: (Option[Either[EE0, AA]], Fiber2[F, EE1, BB]) => F[E3, C],
+      loser: Fiber2[F, EE1, BB],
+      race: Ref2[F, Int],
+      whenDone: Deferred2[F, E3, C]
+    )(res: Option[Either[EE0, AA]]): F[Nothing, Unit] =
+      race.modify { c =>
+        (if (c > 0) monad.unit else whenDone.done(f(res, loser)) map (_ => ())) -> (c + 1)
+      }.flatten
+
+    for {
+      done <- CD.deferred[E3, C]
+      race <- CD.refSet(0)
+      c <- uninterruptible[E3, C](
+            for {
+              left  <- start(fa1)
+              right <- start(fa2)
+              _     <- start[Nothing, Unit](left.await >>= arbiter(leftDone, right, race, done))
+              _     <- start[Nothing, Unit](right.await >>= arbiter(rightDone, left, race, done))
+              rc    <- onInterrupt(done.await)(left.cancel *> right.cancel)
```

Comment:
This cannot ever be interrupted, because it occurs inside `uninterruptible`, so you have to push this to outside the `uninterruptible` region.